File: fib.go
package main

import (
	"fmt"
	"strconv"
)

// The optimized implementation

func fib(n uint64) uint64 {
	if n == 1 || n == 2 {
		return 1
	}

	prev := uint64(1)
	curr := uint64(1)

	for i := uint64(3); i <= n; i++ {
		sum := prev + curr
		prev = curr
		curr = sum
	}

	return curr
}

func main() {
	complx := uint64(50)
	complxResult := fib(complx)
	strResult := strconv.FormatUint(complxResult, 10)

	fmt.Println(strResult)
}

File: fib.jl
# The optimized implementation

function fib(n::UInt64)
    if n == 1 || n == 2
        return UInt64(1)
    end

    prev = UInt64(1)
    curr = UInt64(1)

    for i = 3:n
        sum = prev + curr
        prev = curr
        curr = sum
    end

    return curr
end

complx = UInt64(50)

println(fib(complx))
